Add bulk tagging feature

carvil opened the pull request:
This PR allows us to:

- search for a collection;
- select which content items we want to tag;
- select which taxons we want to tag those content items with;
- preview a tag migration;
- bulk-tag all the content items.

**Note**: Until the publishing API removes the permissions from the `get_content_items` endpoint, we need to add `view_all` permissions to the `dummyapiuser` in development mode in order to see results from the search.

Trello: https://trello.com/c/9LajM53J/10-epic-bulk-tagging-from-existing-taxonomies-to-the-new-one

Work done by both myself and @mobaig (the rebase erased some of the commit details).

What's next:
- [ ] support for topics and browse pages
- [ ] tidy up navigation (perhaps merge some functionality into a single navigation item).

These will come in separate PRs.

carvil commented:
Screenshots to follow shortly

carvil commented:
The flow is:

- Go to the Bulk Tagging page:

<img width="1437" alt="screen shot 2016-09-06 at 17 02 32" src="https://cloud.githubusercontent.com/assets/416701/18281162/c9b1cee2-7453-11e6-907f-3e8dc6d78a7d.png">

- Search for a collection:

<img width="1436" alt="screen shot 2016-09-06 at 17 03 41" src="https://cloud.githubusercontent.com/assets/416701/18281204/e8e4ff28-7453-11e6-89b3-9cde8be58800.png">

- Select which collection to use and then pick the taxons and content items to bulk tag:

<img width="1437" alt="screen shot 2016-09-06 at 17 04 44" src="https://cloud.githubusercontent.com/assets/416701/18281238/0f5f13f0-7454-11e6-91ca-8bdcd4394988.png">

- Preview your new bulk tagging import:

<img width="1437" alt="screen shot 2016-09-06 at 17 06 18" src="https://cloud.githubusercontent.com/assets/416701/18281299/44fc6166-7454-11e6-9d39-5e32bd7f23cd.png">

- Create the tags:

<img width="1434" alt="screen shot 2016-09-06 at 17 06 44" src="https://cloud.githubusercontent.com/assets/416701/18281317/541ecdfa-7454-11e6-9b82-7cc6cb34d0c9.png">

tijmenb commented on app/controllers/bulk_tagging/update_tags_controller.rb:
```diff
@@ -0,0 +1,49 @@
+      params[:content_base_paths]
+    end
+
+    # TODO: make this more readable, weird pattern
```
Could this validation be moved into `BulkTagging::BuildTagMigration`? It looks like a form object anyway.

carvil replied:
Yes, that would make more sense. i will move that into the service class instead.

tijmenb commented on app/controllers/bulk_tagging/collections_controller.rb:
```diff
@@ -0,0 +1,15 @@
+module BulkTagging
```
Would it be clearer for now to merge this and `UpdateTagsController`? Feels like there would be lots of small controllers if we specify one per feature (topics/browse/search etc).

carvil replied:
So originally it was like that, but it makes the routes more complicated and less readable (custom get/post entries with custom paths etc).

By using the conventional way in rails of having a resource and a separate controller we make the routes more readable (at the expense of having more controller classes). But that could be good, otherwise the original controller will be quite big after we are done with the bulk tagging.

What do you think?

tijmenb commented on config/locales/en.yml:
```diff
@@ -26,11 +26,16 @@ en:
-        alert: It was not possible to delete the taxon
+  controllers:
+    bulk_taggings:
+      too_many_results: Your search returned too many results. We are only displaying a subset, please try norrowing down your search.
```
* narrowing

carvil replied:
Whoops!

tijmenb commented on app/controllers/bulk_taggings_controller.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def search
+    if search_response.successful?
```
What kind of errors do we expect here? Perhaps we should let the error bubble up.

carvil replied:
If the API does not respond properly (i.e. with code 200). I'm imagining a timeout (we saw some locally). If we don't do this we get a 500 error...

tijmenb commented on app/models/search_response.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def successful?
+    gds_api_response.code == 200
```
I don't think this is needed, api-adapters will raise an error for all 4XX and 5XX errors anyway.

carvil replied:
Does that mean we would get an exception?

carvil replied:
Talked IRL, we can delete this branch, no need to check for errors as exceptions are raised.

tijmenb commented on app/views/bulk_taggings/new.html.erb:
```diff
@@ -0,0 +1,38 @@
+  <div class="col-md-4">
+    <%= simple_form_for :collection_search, url: search_bulk_tagging_path , method: :get do |f| %>
+      <div class="form-group">
+        <%= f.input :query, input_html: { type: :text, class: 'form-control' }, label: 'Query' %>
```
Should have the current query pre-filled.

carvil replied:
👍

tijmenb commented on app/controllers/bulk_taggings_controller.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def warning_about_multiple_pages
+    if search_response.multiple_pages?
+      flash[:warning] = I18n.t('controllers.bulk_taggings.too_many_results')
```
Don't forget some comment in readme to explain why too much is too much.

carvil replied:
Ok, I will add an entry regarding the bulk tagger to the README.

tijmenb commented on app/views/bulk_tagging/collections/show.html.erb:
```diff
@@ -0,0 +1,42 @@
+          <% expanded_links.each do |expanded_link| %>
+            <tr>
+              <td><%= check_box_tag 'content_base_paths[]', expanded_link.base_path, true %></td>
+              <td><%= expanded_link.title %></td>
```
Perhaps add link to web url as well.

carvil replied:
Yup, makes sense. I will add that link.

tijmenb commented:
From IRL conversation:

- Perhaps replace `original_link_content_id` with a text explanation of the original search

For follow up:

- Let's talk about the naming of all these things
- We need to look at the UI of the collections-tag page (make it different from search)
- Merge tags in index bulk views
- "Select all" checkbox for collections

tijmenb commented on app/controllers/bulk_tagging/collections_controller.rb:
```diff
@@ -0,0 +1,15 @@
+module BulkTagging
+  class CollectionsController < ApplicationController
```
I thought this referred to the `document_collection` format - perhaps rename to something else (don't know what).

carvil replied:
I will move this to a new ticket. I think it makes sense to reconsider the name when we add topics/browse pages too. We would be in a better place to rename this by then.

tijmenb commented on Gemfile:
```diff
@@ -9,7 +9,7 @@ gem 'plek', '~> 1.10'
 gem 'govuk_admin_template', '~> 4.1.1'
 gem 'simple_form', '~> 3.2', '>= 3.2.1'
-gem 'gds-api-adapters', '~> 32.3'
+gem 'gds-api-adapters'
```
Pinning should stay here so we can do `bundle update` and not break.

carvil replied:
Ok, will do.

tijmenb commented on app/services/bulk_tagging/build_tag_mapping.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+
+    def perform
+      TagMapping.new(
```
I'll investigate why this doesn't raise rubocop complaints.

carvil replied:
I will look into this in a separate PR if that's ok.

tijmenb commented on config/routes.rb:
```diff
@@ -19,6 +19,21 @@
     get  'import_progress'
   end
 
+  resources :tag_migrations, except: %i(update edit new) do
```
Can we flip this to `only`? Easier to read the whole routes.

carvil replied:
👍

mobaig commented on app/views/tag_migrations/show.html.erb:
```diff
@@ -0,0 +1,14 @@
+<%= display_header title: 'Preview import', breadcrumbs: [:tag_migrations, "Preview"] do %>
```
Just a note/reminder, possibly this'll get picked up when we revisit naming of things - we should make this header more generic (it's a migration not an import) or it should be passed in from the controller.

carvil replied:
Yes, I will add that and Tijmen's suggestion to that trello card. I think we should revisit naming throughout.

tijmenb commented:
I finally thought of what makes "migration" seem like a wrong word for this. Migration is the moving of things from one thing to another thing. Since we are keeping the original (document collection in this) case, it's not a "migration".

carvil commented:
@tijmenb I've addressed most of the issues you mentioned in the PR. 

I've also added a bunch of follow up tasks to Trello:
- https://trello.com/c/xUoBfOoT/150-bulk-tagging-search-for-topics-and-browse-pages-besides-collections
- https://trello.com/c/Erx6NIwe/151-content-tagger-tidy-up-navigation-in-order-to-consolidate-all-tagging-actions-into-one-navigation-entry
- https://trello.com/c/H403aTpk/152-bulk-tagging-add-description-field-to-tag-migration
- https://trello.com/c/bw6RqXta/153-bulk-tagging-make-sure-the-bulk-taggings-new-and-collections-show-views-are-different-from-each-other-to-avoid-confusion
- https://trello.com/c/lnE4VdSj/154-bulk-tagging-in-the-preview-of-tag-migrations-we-should-consolidate-tag-mapping-records-in-order-to-see-one-row-per-content-item
- https://trello.com/c/IJoqzkbq/155-bulk-tagging-add-select-all-checkbox-that-selects-deselects-all-checkboxes-in-collections-show
- https://trello.com/c/35hTnnQk/156-content-tagger-naming-all-the-things


@mobaig @tijmenb I think we are now in a state where we can merge this and carry on with the tasks above. This PR is already too big and merging and breaking tasks apart would allow all the devs to work on it. WDYT?

tijmenb commented:
Yeah, let's merge this now and fix stuff in follow ups.
